File: password-generator.py
import random
import string

def generate_password(length=12, use_uppercase=True, use_numbers=True, use_special_chars=True):
    if length < 4:
        raise ValueError("Password length should be at least 4.")

    lowercase = string.ascii_lowercase
    uppercase = string.ascii_uppercase if use_uppercase else ''
    numbers = string.digits if use_numbers else ''
    special_chars = "!@#$%^&*()-_=+[]{}|;:,.<>?" if use_special_chars else ''

    all_chars = lowercase + uppercase + numbers + special_chars

    if not all_chars:
        raise ValueError("At least one character set must be enabled.")

    password = [
        random.choice(lowercase),
        random.choice(uppercase) if use_uppercase else random.choice(lowercase),
        random.choice(numbers) if use_numbers else random.choice(lowercase),
        random.choice(special_chars) if use_special_chars else random.choice(lowercase)
    ]

    password += random.choices(all_chars, k=length - len(password))
    random.shuffle(password)
    
    return ''.join(password)

if __name__ == "__main__":
    print("Generated Password:", generate_password(length=16))

File: password-generator.js
function generatePassword(
  length = 12,
  useUppercase = true,
  useNumbers = true,
  useSpecialChars = true
) {
  if (length < 4) {
    throw new Error("Password length should be at least 4.");
  }

  const lowercase = "abcdefghijklmnopqrstuvwxyz";
  const uppercase = useUppercase ? "ABCDEFGHIJKLMNOPQRSTUVWXYZ" : "";
  const numbers = useNumbers ? "0123456789" : "";
  const specialChars = useSpecialChars ? "!@#$%^&*()-_=+[]{}|;:,.<>?" : "";

  const allChars = lowercase + uppercase + numbers + specialChars;

  if (!allChars) {
    throw new Error("At least one character set must be enabled.");
  }

  let password = [
    lowercase[Math.floor(Math.random() * lowercase.length)],
    useUppercase
      ? uppercase[Math.floor(Math.random() * uppercase.length)]
      : lowercase[Math.floor(Math.random() * lowercase.length)],
    useNumbers
      ? numbers[Math.floor(Math.random() * numbers.length)]
      : lowercase[Math.floor(Math.random() * lowercase.length)],
    useSpecialChars
      ? specialChars[Math.floor(Math.random() * specialChars.length)]
      : lowercase[Math.floor(Math.random() * lowercase.length)],
  ];

  for (let i = password.length; i < length; i++) {
    password.push(allChars[Math.floor(Math.random() * allChars.length)]);
  }

  return password.sort(() => Math.random() - 0.5).join("");
}

console.log("Generated Password:", generatePassword(16));
